pymc3/step_methods/smc.py

```diff
@@ -104,599 +59,193 @@ class SMC(atext.ArrayStepSharedLLK):
+            deltas = np.squeeze(proposal(step.n_steps) * step.scaling)
+            for n_step in range(0, step.n_steps):
+                delta = deltas[n_step]
+
+                if any_discrete:
+                    if all_discrete:
+                        delta = np.round(delta, 0).astype('int64')
+                        q_old = q_old.astype('int64')
+                        q_new = (q_old + delta).astype('int64')
+                    else:
+                        delta[discrete] = np.round(delta[discrete], 0)
+                        q_new = (q_old + delta)
                 else:
-                    x_chains = None
-            else:
-                step.covariance = step.calc_covariance()
-                step.proposal_dist = choose_proposal(step.proposal_name, scale=step.covariance)
-                step.resampling_indexes = step.resample(chains)
-                step.chain_previous_lpoint = step.get_chain_previous_lpoint(mtrace, chains)
+                    q_new = floatX(q_old + delta)
 
-                stage_handler.dump_atmip_params(step)
+                new_tempered_post = prior_logp(*q_new) + likelihood_logp(*q_new) * beta
 
-                step.stage += 1
+                q_old, accept = metrop_select(new_tempered_post - old_tempered_post, q_new, q_old)
+                if accept:
+                    accepted += accept
+                    posterior[draw] = q_old
+                    old_tempered_post = new_tempered_post
+                proposed += 1.
 
-        # Metropolis sampling final stage
-        pm._log.info('Sample final stage')
-        step.stage = -1
-        x_chains = stage_handler.clean_directory(step.stage, x_chains, rm_flag)
-        weights_un = np.exp((1 - step.old_beta) * (step.likelihoods - step.likelihoods.max()))
-        step.weights = weights_un / np.sum(weights_un)
-        step.covariance = step.calc_covariance()
-        step.proposal_dist = choose_proposal(step.proposal_name, scale=step.covariance)
-        step.resampling_indexes = step.resample(chains)
-        step.chain_previous_lpoint = step.get_chain_previous_lpoint(mtrace, chains)
+        acc_rate = accepted / proposed
+        stage += 1
 
-        x_chains = nr.randint(0, chains, size=samples)
+    trace = _posterior_to_trace(posterior, model)
 
-        sample_args['draws'] = step.n_steps_final
-        sample_args['step'] = step
-        sample_args['stage_path'] = stage_handler.stage_path(step.stage)
-        sample_args['x_chains'] = x_chains
-        _iter_parallel_chains(**sample_args)
+    return trace
 
-        stage_handler.dump_atmip_params(step)
-
-        #model.marginal_likelihood = step.sjs
-        return stage_handler.create_result_trace(step.stage,
-                                                 step=step,
-                                                 model=model)
-
-
-def _initial_population(samples, chains, model, variables):
+# FIXME!!!!
+def _initial_population(samples, model, variables):
```

Comment:
You can use the sample_prior_predictive function here right?

Reply:
That's an option. Currently I need an array of samples not a dictionary, but this is something I need to change in order to work with multidimensional variables.